### ISSUE TITLE
parallel to_sql

### ISSUE THREAD

**Issue by eavidan** [modin-project/modin#453]
Today there are many ways to read into modin in parallel, but none of the write (`to_`) methods are implemented. Since `to_sql` does not require parallel file access, it could be a good candidate for the first parallel write.
It should be fairly easy to implement. following is a simple implementation suggestion. would appreciate your inputs

most of the implementation is in `pandas_query_compiler.py`
```py
    def to_sql(self, **kwargs):
        kwargs["index"] = False
        kwargs["index_label"] = None
       # this is to create the full table and can be used for validation of the input as well. making the alignment with pandas easier
        empty = self.head(0).to_pandas() 
        empty.to_sql(**kwargs)
        # so each partition will append its respective DF
        kwargs["if_exists"] = "append"
        columns = self.columns  

        def func(df, **kwargs):
            df.columns = columns # get the column names to the partition
            df.to_sql(**kwargs)

        map_func = self._prepare_method(func, **kwargs)
        self.map_across_full_axis(0, map_func)
```

Though we have an empty INSERT at the beginning and a shuffle, the improvement in performance should still be definitive

**Comment by devin-petersohn** [comment]
Thanks @eavidan this sounds great!

In the code example, mapping across `axis=1` might be what you want as it will be a full row for each insert. I am interested to see if there is any reordering (e.g. the rows in the same order as the dataframe).

**Comment by eavidan** [comment]
@devin-petersohn thats a good point. I was more focused on the columns, and did not think about the rows. 

After some testing, that is indeed the case, but we should not really care about that as SQL tables do not maintain row order by definition. So, pandas will not keep the row order as well.
Usually the way you read in the same order is by saving the index using `df.to_sql(index=True)`. 
Then you can read with `pandas.read_sql(index_col="index")` and get the original ordering.
The thing is that we don't have the original index in each partition, and the user will have a table with a "misleading" index column in the DB.

Any suggestions on how we can keep the original row index in each partition?

**Comment by eavidan** [comment]
so for now I believe the easiest way is adding an index column to the DF it self before the operation and removing it at the end. Not sure how much it will hurt performance, since its 3 distributed operations, though those two (add and remove) should be really light.
any thoughts?

**Comment by devin-petersohn** [comment]
Hey @eavidan, that sounds great. I will offer an alternative that is very similar.

One option is to do a `df.reset_index(drop=False)` first. Then you can do row-wise inserts with `df.to_sql(index=False)` in the partitions (because the data will already be there). It is similar to your proposal, but we already have `reset_index` implemented. You can do `df.index.name = index_label` first to get the index column named properly.

It is very similar to your approach, but your approach may be a bit more direct. Either way sounds good to me!

**Comment by eavidan** [comment]
thanks @devin-petersohn 
one last thing before the PR. I have an initial working implementation, but for some reason I am getting an error. The implementation is in this [branch].(https://github.com/eavidan/modin/tree/issue_453)
But, when I run the code below, I get this error and an empty DF back from the DB: 
```
Traceback (most recent call last):
  File "/Users/eavidan/PycharmProjects/forks/modin/modin/engines/ray/pandas_on_ray/axis_partition.py", line 74, in deploy_ray_func
    return func(*args)
  File "/Users/eavidan/PycharmProjects/forks/modin/modin/engines/base/axis_partition.py", line 208, in deploy_axis_func
    if sum(lengths) != len(result.columns):
AttributeError: 'NoneType' object has no attribute 'columns'
```
It seems that I am missing some blocking operation, since as I put `time.sleep(2)`, the DF is inserted in full to the DB. The error is thrown anyway at the beginning. I could be completely off of course.
What am I missing here?

```py
import modin.pandas as pd
import pandas
import os
import numpy as np
import time

conn = "sqlite:///tmp.db"

if os.path.exists("tmp.db"):
    os.remove("tmp.db")
df = pd.DataFrame(np.random.randint(0, 100, size=(100, 100)), columns=range(100))

df.to_sql(name="tbl", con=conn)

# time.sleep(2)

df = pandas.read_sql("tbl", con=conn, index_col="index")
print(df)
```

**Comment by devin-petersohn** [comment]
Ok I found the source of the problem. It is as you said: the operation is non-blocking so the read is happening before the writes are finished (or even started in some cases).

One of the issues I found was that the internal function is returning `None` values, but then the code is trying to create a new `PandasQueryCompiler` with those `None`s.

I will PR against your branch for the simple fix I created, though feel free to modify it. I also copied your logic into the `pandas_on_ray.io.PandasOnRayIO` class because there we can use ray specific optimizations for I/O.
